subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformEnvironmentIntegrationTest.groovy

```diff
@@ -0,0 +1,242 @@
+    private static final JUNIT_JUPITER_VERSION = '5.8.1'
+    private static final JUNIT_PLATFORM_VERSION = '1.8.1'
+    private static final OPENTEST4J_VERSION = '1.2.0'
+
+    def setup() {
+        buildFile << """
+            plugins {
+                id 'java-library'
+            }
+
+            ${mavenCentralRepository()}
+
+            test {
+                useJUnitPlatform()
+
+                systemProperties.isJava9 = "\${JavaVersion.current().isJava9Compatible()}"
+                systemProperties.testSysProperty = 'value'
+                systemProperties.projectDir = projectDir
+                environment.TEST_ENV_VAR = 'value'
+                testLogging.showStandardStreams = true
+            }
+        """
+    }
+
+    def "should prompt user to add dependencies when they are not in test runtime classpath"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                compileOnly 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+            }
+        """
+        file('src/test/java/org/example/ExampleTest.java') << """
+            package org.example;
+            public class ExampleTest {
+                @org.junit.jupiter.api.Test
+                public void ok() { }
+            }
+        """
+
+        when:
+        fails('test')
+
+        then:
+        new DefaultTestExecutionResult(testDirectory)
+            .testClassStartsWith('Gradle Test Executor')
+            .assertExecutionFailedWithCause(containsString('consider adding an engine implementation JAR to the classpath'))
+    }
+
+    def "automatically loads junit-platform-launcher from distribution"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                implementation 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+            }
+        """
+
+        addClasspathTest("""
+            assertEquals(jars.get(0), "junit-jupiter-params-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(1), "junit-jupiter-engine-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(2), "junit-jupiter-api-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(3), "junit-platform-engine-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(4), "junit-platform-commons-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(5), "junit-jupiter-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(6), "opentest4j-${OPENTEST4J_VERSION}.jar");
+
+            // The distribtuion-loaded jar can vary in version, since the GE test acceleration
+            // plugins inject their own versions of these libraries during integration tests.
+            // See: https://github.com/gradle/gradle/pull/21494
+            assertTrue(jars.get(7).startsWith("junit-platform-launcher-1."));
+
+            assertEquals(jars.size(), 8);
```

Comment:
❔ This seems unnecessary, why care how many transitive deps junit pulls in here?

Reply:
We need to verify that there are no extra dependencies that we aren't missing here. If for example we accidentally load one of the `engine` implementations what was pulled in from the test runtime classpath.